src/magic_di/healthcheck.py

```diff
@@ -0,0 +1,48 @@
+import asyncio
+from asyncio import Future
+from dataclasses import dataclass
+from typing import Any, Protocol
+
+from magic_di import Connectable, ConnectableProtocol, DependencyInjector
+
+
+class PingableProtocol(ConnectableProtocol, Protocol):
+    async def __ping__(self) -> None: ...
+
+
+@dataclass
+class DependenciesHealthcheck(Connectable):
+    """
+    Injectable Healthcheck component that pings all injected dependencies
+    that implement the PingableProtocol
+
+    Example usage:
+
+    ```python
+    from app.components.services.health import DependenciesHealthcheck
+
+    async def main(redis: Redis, deps_healthcheck: DependenciesHealthcheck) -> None:
+        await deps_healthcheck.ping_dependencies()  # redis will be pinged if it has method __ping__
+
+    inject_and_run(main)
+    ```
+    """
+
+    injector: DependencyInjector
+
+    async def ping_dependencies(self, max_concurrency: int = 1) -> None:
```

Comment:
Would it be beneficial if this would return information about which ones failed (and which one succeeded)? Then it would be possible to include that information in the response of the healthcheck endpoint. At minimum, I think there could be some logging inside this method about failures.

Reply:
Currently in daas-core we are using this approach: each dependency (such as redis, mongo or kafka-consumer) implements [a ping method](https://github.com/creditornot/daas-core/blob/2ce230fb6df47748180652a798b22246dd58808b/app/kafka/consumer.py#L127-L140), that might raise an exception. Each of those pings has they own set of exceptions (e.g., `ConnectionFailure`, `NetworkTimeout`, `OperationFailure` for MongoDB), and we’ve intentionally decided not to catch them.

This approach ensures that in case of fire, we get a complete stack trace in the logs, making it easier to debug what happened. 

And it also leads to 500 error status code for a [typical FastAPI/Starlette server](https://github.com/creditornot/daas-core/blob/ac0b2c25d2808b8b1865fdc1b3e866a9e2c8fc9e/app/common/healthcheck_server.py#L27-L29). Since these health checks are primarily designed for Kubernetes - correct status code is basically [the only thing that matters](https://github.com/kubernetes/kubernetes/blob/67f72d0e90f8a57c7e761a5fb34f6c4b21c25462/pkg/probe/http/http.go#L111)

What do you think about it?

Reply:
> Would it be beneficial if this would return information about which ones failed (and which one succeeded)?

Hmm. Do we have any case where we need to know which dependencies are healthy or not?
Usually we only want to know if everything is healthy.

> At minimum, I think there could be some logging inside this method about failures.

That's good point, yeah. I added debug logging. But I decided not to log failures since its raised to user and user can catch it and log if needed.

But, what I’m thinking about is, do we need the ﻿ignore_exceptions argument? To ignore some concrete exceptions that are raised by some dependencies

Reply:
Yeah, I guess it makes sense, at least in K8s (probes) use cases. I was just thinking that peeps in the wider community might have also non-K8s use cases in which it might be beneficial to know which ones failed. Anyway, I guess this will be sufficient for majority.

> do we need the ﻿ignore_exceptions argument
At least I can't think of a real world use case in which one would like to check the healthiness of all dependencies but ignore some 🤔